paradox-nakadi-consumer-partitioned-zk/src/main/java/de/zalando/paradox/nakadi/consumer/partitioned/zk/ZKConsumerLeader.java

```diff
@@ -54,151 +47,69 @@
-                            switch (connectionState) {
-
-                                case LOST :
-                                case SUSPENDED :
-                                    leaderControl.relinquishLeadership();
-                                    throw new CancelLeadershipException("State " + connectionState + ":" + t);
-
-                                default :
-                                    break;
-                            }
-                        }
+                        super.stateChanged(client, connectionState);
                     }
 
                 });
         selector.setId(member.getMemberId());
 
-        if (null == leaderControls.putIfAbsent(t, LeaderControl.of(selector))) {
+        if (null == leaderSelectors.putIfAbsent(t, selector)) {
             LOGGER.info("Init member [{}] leadership for [{}]", member.getMemberId(), t);
 
             // restart selector every time the leadership is lost
+            selector.autoRequeue();
+
             try {
                 selector.start();
             } catch (final Throwable throwable) {
-                leaderControls.remove(t);
+                leaderSelectors.remove(t);
+                selector.close();
                 ThrowableUtils.throwException(throwable);
             }
+        } else {
+            selector.close();
         }
     }
 
-    public Map<String, Boolean> getParticipants(final T t) {
-        return Optional.ofNullable(leaderControls.get(t)).map((leaderControl) ->
-                               leaderControl.getParticipants().stream().collect(
-                                   Collectors.toMap(Participant::getId, Participant::isLeader, (u, v) -> u))).orElseGet(
-                           Collections::emptyMap);
-    }
-
     public void closeGroupLeadership(final T t) {
-        final LeaderControl leaderControl = leaderControls.remove(t);
-        if (null != leaderControl) {
+        final LeaderSelector leaderSelector = leaderSelectors.remove(t);
+        if (null != leaderSelector) {
             LOGGER.info("Close member [{}] leadership for [{}]", member.getMemberId(), t);
-            leaderControl.relinquishLeadership();
+            CloseableUtils.closeQuietly(leaderSelector);
         }
     }
 
     public void close() {
         LOGGER.info("Closing for member [{}]", member.getMemberId());
-
-        leaderControls.values().forEach(LeaderControl::relinquishLeadership);
-
-        try {
-            Thread.sleep(1000L);
-        } catch (InterruptedException e) {
-            ThrowableUtils.throwException(e);
-        }
-
-        leaderControls.values().forEach(leaderControl -> {
-            try {
-                leaderControl.selector.close();
-            } catch (Exception e) {
-                LOGGER.warn("Unexpected error while closing leader selector [{}]", e);
-            }
-        });
-        leaderControls.clear();
+        leaderSelectors.values().forEach(CloseableUtils::closeQuietly);
```

Comment:
Is this thread safe? Could another leaderSelector be added while the close is processed but before clear was executed?